wip-trampoline.md

```diff
@@ -0,0 +1,138 @@
+# Trampoline Onion Routing
+
+This document describes the additions we make to the existing onion routing
+to enable the use of trampoline hops.
+
+## Packet Structure
+
+### Trampoline Onion
+
+We define a fixed-size trampoline onion packet that can be nested inside onion
+packets. We define `N` to be the maximum number of trampoline nodes (excluding
+the sender node and the destination node).
+
+The trampoline packet always contains `2*N*FRAME_SIZE + 66` bytes.
+
+1. type: `trampoline_onion_packet`
+2. data:
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`2*N*FRAME_SIZE`:`trampoline_hops_data`]
+   * [`32`:`hmac`]
+
+Note: we use the same onion format as the standard onion which wastes frames,
+but that theoretically allows another level of nesting (mind blown).
+However not really usable given the space constraints...
+
+TODO: maybe this should be version 1 of the overall onion version namespace?
+
+TODO: maybe we can do something more clever with TLV? At the cost of a second
+onion parser (trampoline onion parser)? Use the top-level version in the parser
+to discriminate?
+
+TODO: version maybe redundant with the TLV type that will include this?
+But keeping it allows using this packet to be used as a top-level onion packet
+where we basically let each hop choose which channel to use to route to the
+next hop. Is this useful?
+
+The `trampoline_hops_data` field is a list of `trampoline_hop_payload`s.
+A `trampoline_hop_payload` is a structure that holds obfuscations of the next
+trampoline hop's address, transfer information, and its associated HMAC.
+
+It is `2*N*FRAME_SIZE` long and has the following structure (where `n1` and
+`n2` are the number of frames used by trampoline hops `1` and `2`):
+
+1. type: `trampoline_hops_data`
+2. data:
+   * [`n1*FRAME_SIZE`:`trampoline_hop_payload`]
+   * [`n2*FRAME_SIZE`:`trampoline_hop_payload`]
+   * ...
+   * `filler`
+
+Each `trampoline_hop_payload` has the following structure:
+
+1. type: `trampoline_hop_payload`
+2. data:
+   * [`1`:`num_frames_and_realm`]
+   * [`1`:`type`]
+   * [`1`:`length`]
+   * [`33`:`node_id`]
+   * [`8`:`amt_to_forward`]
+   * [`4`:`outgoing_cltv_value`]
+   * `optional TLV can be added here`
+   * `padding`
+   * [`32`:`HMAC`]
+
+For this specification, some restrictions apply:
+
+* `realm` MUST be `0x01` to indicate that the payload contains a TLV stream
+and `num_frames` MUST correctly reflect the number of frames used.
+* the TLV stream MUST contain the `type` `0x01` (`next_trampoline_hop`)
+* `next_trampoline_hop`'s length MUST be `33+8+4=45` bytes
+* the TLV stream MAY contain other optional types
+* `padding` MUST use only `0x00` bytes
```

Comment:
I'd prefer sticking with a raw TLV on the top-level that is easier to change later, rather than rebuilding the legacy `hop_data_v0` format in this new layer.

Reply:
I agree, I'll clarify this while removing the fixed frames entirely.